### ISSUE TITLE
Readiness notification

### ISSUE THREAD

**Issue by WhyNotHugo** [dbus2/busd#27]
## Feature request

Readiness notification allows a daemon to notify the service manager when it has finalised its start-up sequence. This is very important so the service manager can handle dependencies properly.

In particular, when a service depends on dbus, it needs to be started after the broker has started listening on the configured address, and not merely after the dbus broker process has started.

## Prior art

[s6 has a very well though out (yet also extremely simple) readiness notificationmechanism](https://www.skarnet.org/software/s6/notifywhenup.html): the service manager hands over a file descriptor to a service, and the service is expected write `\n` to indicate that it is up and listening.

If the file descriptor is closed without anything being written to it, it is assumed that the service filed to start properly.

systemd uses a slightly different approach, and passes a socket where it expects the service to write `READY=1\n` when it is ready to accept connections (it also reuses this channel for other, unrelated communication).

The `swaylock` developers have figured out an approach that works well with both s6 and system: https://github.com/swaywm/swaylock/pull/281. In essence, the service takes a file descriptor (parametrised via `--ready-fd N`), writes `READY=1\n` into it, and then closes it. This fulfils the expectation for both systemd and s6. (in the case of `swaylock`, it is common to want to suspend the system _after_ it has been locked, so this is also used for ordering dependencies).

To summarise the idea: `busd` could be started with something like `busd --ready-fd 3` and it would write `READY=1\n` to file descriptor 3 when it is ready to receive incoming connections, indicting to the service manager that dependant services can now be started.

**Comment by WhyNotHugo** [comment]
I'm opening a ticket rather than jumping directly into hacking since I'm sure this warrants some discussion first :)

**Comment by zeenix** [comment]
@WhyNotHugo Thanks for filing this. Since we setup the socket almost immediately, I wonder if this really is a problem? :thinking: The service manager can simple wait for the socket file to appear?

**Comment by zeenix** [comment]
> To summarise the idea: `busd` could be started with something like `busd --ready-fd 3` and it would write `READY=1\n` to file descriptor 3 when it is ready to receive incoming connections, indicting to the service manager that dependant services can now be started.

But sure, if we are sure this is actually needed, this idea seems good to me.

**Comment by WhyNotHugo** [comment]
> Thanks for filing this. Since we setup the socket almost immediately, I wonder if this really is a problem?

If there is not readiness notification, a service manager will start dependant services as soon as the `busd` process starts. These will also try to connect to the socket almost immediately too, so it ends up being a recipe for race conditions.

> The service manager can simple wait for the socket file to appear?

That would require a dedicated helper. It' perfectly feasible, but it's always simper if the service itself can indicate "I'm ready" instead of having a separate process checking if the socket is there.

**Comment by WhyNotHugo** [comment]
For what it's worth, I think `dbus-daemon` guarantees that `--print-pid` prints after it has begun listening to a socket, so it can also be used as a readiness notification mechanism.

**Comment by zeenix** [comment]
> For what it's worth, I think `dbus-daemon` guarantees that `--print-pid` prints after it has begun listening to a socket, so it can also be used as a readiness notification mechanism.

Sure, why not.
